docs/enhancements/clusterpool-inventory.md

```diff
@@ -0,0 +1,175 @@
+      message: patch is currently in use by cluster deployment foo of clusterpool foo-pool
+      reason: PatchInUse
+      status: "false"
+      type: PatchAvailable
+```
+
+if DNS is configured with the name `foo`, The base 64 encoded json content to patch vSphere install config will be as follows
+
+```json
+[
+  { "op": "replace", "path": "metadata/name", "value": "foo" }
+]
+```
+
+When adding a ClusterDeployment, if such an `Inventory` is present, ClusterPool controller will:
+- Load up the inventory list.
+- For each reference it will do a GET to fetch InstallConfigPatch and check the if the PatchAvailable condition is true.
+- For the first InstallConfigPatch that is available, it will use the json patch in the `spec.patch` field to update the default install config generated by clusterpool. It will also update the status with reference to the cluster deployment that is using it and update the `PatchAvailable` condition to false.
+- Set the `spec.clusterPoolReference.installConfigPatchRef`field in the ClusterDeployment with a reference to InstallConfigPatch CR.
+
+Absent an `Inventory`, ClusterPool will continue to use the generated default install config as it does today.
+
+When ClusterDeployment is deleted, ClusterPool controller will: 
+- Find out if ClusterDeployment used an InstallConfigPatch from `spec.ClusterPoolReference.InstallConfigPatchRef`field.
+- If yes, we simply update InstallConfigPatch, setting the `spec.clusterDeploymentRef` to nil and `PatchAvailable` to true.  
+
+### How To Use
+For the VSphere case, this allows the administrator to:
+- Preconfigure DNS with following entries (assuming cluster name is `foo`)
+    ```
+    10.0.0.10  api.foo.example.com
+    10.0.0.11  apps.foo.example.com
+    10.0.0.12  console.foo.example.com
+    ```
+- Create a InstallConfigPatch CR to patch `spec.metadata.name` field of the default install config generated by clusterpool controller. Please refer the section above of a sample CR. The base64 encoded json content in `spec.patch` field should be as follows
+    ```json
+    [
+      { "op": "replace", "path": "metadata/name", "value": "foo" }
+    ]
+    ```
+- Add the name of InstallConfigPatch CR to`clusterPool.spec.inventory.installConfigPatches` list. For InstallConfigPatch with a name `foo-install-config-patch` the clusterpool should be configured as follows
+    ```yaml
+    spec:
+      inventory:
+        installConfigPatches:
+          - name: foo-install-config-patch
+    ```
+  
+### Validation
+Webhook validation will ensure that for a clusterpool
+- if `spec.inventory.installConfigPatchs` is specified, it is not an empty list.
+- if `spec.inventory.installConfigPatchs` is specified, `spec.maxSize` is equal to or less than length of list `spec.inventory.installConfigPatchs`.
+- if `spec.inventory.installConfigPatchs` is specified, `spec.size` is equal to or less than length of list `spec.inventory.installConfigPatchs`.
+
```

Comment:
Other validation scenarios to consider:
- At what point do we validate that an inventory entry has an InstallConfigPatch associated with it? What do we do if it doesn't? (I think we do this in the flow where we're trying to create a new CD; and I think we log an error and move on to the next one.)
- What do we do if a patch fails, e.g. because it's malformed? (I think we log an error, add a status condition to the InstallConfigPatch, and move on to the next one. But the status condition should be write-only -- e.g. we shouldn't subsequently ignore the patch, because we can't know whether, when, or how the user has "fixed" it.)

Reply:
Good points! Both your suggestions are fair enough, we would also like to add degraded condition on Clusterpool to notify problems with InstallConfigPatches